### ISSUE TITLE
Revisit #20883 before release

### ISSUE THREAD

**Issue by thockin** [kubernetes/kubernetes#20930]
Should we enable or disable hairpin?

ref #20883

**Comment by bprashanth** [comment]
For reference it's on by default today (so kubelet sets hairpin flag)

**Comment by thockin** [comment]
For reference it is OFF in some e2e tests.

**Comment by AdoHe** [comment]
@thockin so we just add a flag to control this at last?

**Comment by bgrant0607** [comment]
@thockin Please self-triage (team/priority labels)

**Comment by bprashanth** [comment]
I think we need to figure out the default userspace vs kernel proxy issue for 1.2. For that we're waiting on soak test to run its course. Assigning to myself. 

Fyi current winning solution seems to be: disable it (to avoid kernel bug), document impact and how one installs a watchdog rebooter (using sysctl), improve perf of userspace proxy.

**Comment by thockin** [comment]
@AdoHe this is a temporary things while we root out the nasty bug.

**Comment by thockin** [comment]
@bprashanth what do you think.  Should we revert this PR, now that we decided on promiscuous?  Is there any chance we might want to be non-promiscuous with this disabled?

**Comment by AdoHe** [comment]
@thockin you mean that we will enable promiscuous mode on the container side interface?

**Comment by bprashanth** [comment]
> Is there any chance we might want to be non-promiscuous with this disabled?

no on the container bridge.

> Is there any chance we might want to be non-promiscuous with this disabled?

@thockin yes they can set hairpin-mode and the recommended sysctl setting by themselves if they don't want promiscuous mode and don't mind a reboot on hung task. I think we're stuck with the flag for now, because it now has promiscous-mode, hairpin-veth and none, the last one meaning do nothing (eg non bridge networks).

**Comment by thockin** [comment]
What I meant is whether we can reduce this to (non-promisc+hairpin ||
promisc).  Do we really think anyone will want non-promis+non-hairpin (i.e.
partially broken)

**Comment by bprashanth** [comment]
Another issue was brought up on https://github.com/kubernetes/kubernetes/issues/20475. There's a segment of users who might've had --configure-cbr0=false (so we won't set promiscusou mode now) AND still be using docker0 and taking advantage of the fact that previously we were always setting the hairpin flag on their veths. 

The new default of "hairpin-mode=off" protects them from the kernel bug. However they will still not have feature compat with previous release. I think this is a small segment, and people setting --configure-cbr0=false will be using something like ovs, but we need to document or address this in some way.

**Comment by thockin** [comment]
OK.  I revisited the final design of this flag, and I guess we should leave it.

As for configure-cbr0=false, I think we should never have been setting hairpin in that case, but we probably need to continue to make it work.  As far as I can see the default is promiscuous bridge, though, not "off".  The logic here is a little twisty.  PR coming, which I hope will clean up a bit.

**Comment by bprashanth** [comment]
> As far as I can see the default is promiscuous bridge, though, not "off". 

Yeah, its promiscusou bridge, but we won't touch the bridge if --configure-cbr0 is false (we don't know which bridge to set it on), so it's "all off" aka "none" if you set: --configure-cbr0=false --hairpin-mode=promiscuous-mode

**Comment by thockin** [comment]
maybe we should default back to hairpin-veth if !CBR0

**Comment by thockin** [comment]
I think we should preserve behavior.  Talk me out of it?

**Comment by thockin** [comment]
It's a little clumsy that the default values are promiscuous and !cbr0, which is basically invalid.  But I don't am not sure that changing the defaults is worth the risk at this point (the risk of missing some salt or something)

**Comment by bprashanth** [comment]
I think the old behavior was wrong to begin with (setting hairpin on all interfaces annoyed some people: https://github.com/kubernetes/kubernetes/pull/21328). I also think a majority of people using --configure-cbr0 will either be using a differnet networking solution like ovs (in which case setting hairpin might actually expose them to the bug for no reason -- doubtful though) or something like flannel (they will be broken). Breaking flannel people will force them to lookup their kernels and understand the implications of hairpin mode.

Eitherway there's a chance of breakage, though in the kernel bug case the breakage is harder and in-your-face. 

>It's a little clumsy that the default values are promiscuous and !cbr0, which is basically invalid. But I don't am not sure that changing the defaults is worth the risk at this point (the risk of missing some salt or something)

Those are great defaults. Why salt? setting if !cbr0 {hairpin-mode=hairpin-veth} should be easy in the kubelet?

**Comment by thockin** [comment]
The old behavior WAS wrong, but configure-cbr0=false is the default and it seems that a lot of people followed instructions that did not involve changing that.  They are already living with the issue, I'd rather them make an explicit choice to switch to promiscuous mode rather than have something that used to work stop working.

**Comment by bprashanth** [comment]
Fair enough, though it stopped working for a bug that is exacerbated by kernel iptables proxy. This is the moral equivalent of us picking "sane defaults".

**Comment by bgrant0607** [comment]
@thockin Time to revisit. I'd like 1.2.0-beta.1 cut Friday morning with everything other than maybe bug fixes. Do we have a decision?

**Comment by bprashanth** [comment]
I believe this is all that was needed https://github.com/kubernetes/kubernetes/pull/22664

**Comment by goltermann** [comment]
@thockin can we close this one?
